File: KNN.py
import pandas as pd
from matplotlib import pyplot as plt
import math

def seprate(t1,t2,t3,data):
    for i in range (0,data['species'].count()):
        if data['species'].values[i]=='setosa':
            t1.append(i)
        elif data['species'].values[i]=='versicolor':
            t2.append(i)
        elif data['species'].values[i]=='virginica':
            t3.append(i)
    return t1,t2,t3

def ready(train,t1,t2,t3):
    id=int(len(t1)*0.7)
    for i in range (0,id):
        train.append(t1[i])
    id=int(len(t2)*0.7)
    for i in range (0,id):
        train.append(t2[i])
    id=int(len(t3)*0.7)
    for i in range (0,id):
        train.append(t3[i])
    return train

def testing(test,t1,t2,t3):
    for i in range (int(len(t1)*0.7),len(t1)):
        test.append(t1[i])
    for i in range (int(len(t2)*0.7),len(t2)):
        test.append(t2[i])
    for i in range (int(len(t3)*0.7),len(t3)):
        test.append(t3[i])
    return test

def predict(data,test,train,k):
    result={}
    correct=0

    for i in range (0,tr):
        sl=data['sepal_length'][test[i]]
        sw=data['sepal_width'][test[i]]
        pl=data['petal_length'][test[i]]
        pw=data['petal_width'][test[i]]
        for j in range(0,len(train)):
            sl2=data['sepal_length'][train[j]]-sl
            sw2=data['sepal_width'][train[j]]-sw
            pl2=data['petal_length'][train[j]]-pl
            pw2=data['petal_width'][train[j]]-pw
            ans=sl2*sl2+sw2*sw2+pl2*pl2+pw2*pw2
            ans=math.sqrt(ans)
            result[ans]=data['species'][train[j]]
        c1=0
        c2=0
        c3=0
        count=0

        for z in sorted(result):
            if(count==k): break
            if(result[z]=='setosa'): c1=c1+1
            if(result[z]=='versicolor'): c2=c2+1
            if(result[z]=='virginica'): c3=c3+1
            count=count+1
        mx=max(c1,max(c2,c3))
        if(mx==c1):
            if(data['species'][test[i]]=="setosa"):
                print(test[i]," yes ",data['species'][test[i]])
                correct=correct+1
            else:
                print(test[i]," no ",data['species'][test[i]])
        elif(mx==c2):
            if(data['species'][test[i]]=="versicolor"):
                print(test[i]," yes ",data['species'][test[i]])
                correct=correct+1
            else:
                print(test[i]," no ",data['species'][test[i]])
        elif(mx==c3):
            if(data['species'][test[i]]=="virginica"):
                print(test[i]," yes ",data['species'][test[i]])
                correct=correct+1
            else:
                print(test[i]," no ",data['species'][test[i]])
        result.clear()
    return correct


data = pd.read_csv('../input/iriscsv/IRIS.csv')
t1=[]
t2=[]
t3=[]

t1,t2,t3=seprate(t1,t2,t3,data)
test=[]
train=[]
train=ready(train,t1,t2,t3)
test=testing(test,t1,t2,t3)
tr=len(test)

nn=[]
acc=[]
for k in range(1,100,2):
    correct=0
    correct=predict(data,test,train,k)
    per=correct/len(test)*100
    print("Accuracy of k = ",k," is ",per)
    nn.append(k)
    acc.append(per)

plt.plot(nn,acc)
plt.xlabel("value of k")
plt.ylabel("accuracy")
plt.title("GRAPH BETWEEN ACCURACY AND K")
plt.show()

File: KNN.cpp
#include<bits/stdc++.h>
using namespace std;

struct first{
    float arr[4];
    int type;
};

bool cmp(pair<int,int> p1, pair<int,int> p2)
{
    return (p1.first < p2.first);
}

//To split data into train dataset and test dataset
void split(vector<first> &train, vector<first> &test)
{
    fstream f;
    f.open("/home/cryptoduh/Downloads/Iris.csv",ios::in);
    string line;
    getline(f,line);
    int d=1;
    first tr,te;

    int n = (70*50)/100;   //To make 70% data as train data of each flower type
    
    //To split Setosa flower type into train data and test data
    while(d != 51 && getline(f,line))
    {
        stringstream s(line);
        string temp;
        if(d < n)
        {
            int i = 0;
            while(getline(s, temp, ','))
            {
                if(i != 4)
                {
                    float result = std::stof(temp);
                    tr.arr[i] = result;
                    i++;
                }
            }
            tr.type = 1;
            train.push_back(tr);
            d++;
        }
        else
        {
            int i = 0;
            while(getline(s, temp, ','))
            {
                if(i != 4)
                {
                    float result = std::stof(temp);
                    te.arr[i] = result;
                    i++;
                }
            }
            te.type = 1;
            test.push_back(te);
            d++;
        }
    }

    //To split Versicolor flowers data into train data and test data
    while(d != 101 && getline(f,line))
    {
        stringstream s(line);
        string temp;
        if(d < 50+n)
        {
            int i = 0;
            while(getline(s, temp, ','))
            {
                if(i != 4)
                {
                    float result = std::stof(temp);
                    tr.arr[i] = result;
                    i++;
                }
            }
            tr.type = 2;
            train.push_back(tr);
            d++;
        }
        else
        {
            int i = 0;
            while(getline(s, temp, ','))
            {
                if(i != 4)
                {
                    float result = std::stof(temp);
                    te.arr[i] = result;
                    i++;
                }
            }
            te.type = 2;
            test.push_back(te);
            d++;
        }
    }

    //To split virginica flower data into train data and test data
    while(d != 151 && getline(f,line))
    {
        stringstream s(line);
        string temp;
        if(d < 100+n)
        {
            int i = 0;
            while(getline(s, temp, ','))
            {
                if(i != 4)
                {
                    float result = std::stof(temp);
                    tr.arr[i] = result;
                    i++;
                }
            }
            tr.type = 3;
            train.push_back(tr);
            d++;
        }
        else
        {
            int i = 0;
            while(getline(s, temp, ','))
            {
                if(i != 4)
                {
                    float result = std::stof(temp);
                    te.arr[i] = result;
                    i++;
                }
            }
            te.type = 3;
            test.push_back(te);
            d++;
        }
    }
}

//Main function used to predict
int KNN(int K, vector<first>&test, vector<first>&train)
{
    int len = test.size();
    int correct = 0;
    int wrong = 0;
    
    for(int i=0; i<len; i++)
    {
        vector< pair<float, int> >g;

        for(int j=0; j<train.size(); j++)
        {
            float d = 0;
            for(int k=0; k<K; k++)
                d += pow((train[j].arr[k]-test[i].arr[k]),2);

            g.push_back(make_pair(sqrt(d),train[j].type));
        }

        sort(g.begin(), g.end(), cmp);        //Sort distance to choose nearest K neighbours

        int s=0,ver=0,vir=0;
        for(int j=0; j<K; j++)
        {
            if(g[j].second == 1)
                s++;
            else if(g[j].second == 2)
                ver++;
            else 
                vir++;
        }

        int max;
        if(s >= ver && s >= vir)
            max = 1;
        else if(ver >= s && ver >= vir)
            max = 2;
        else 
            max = 3;
        
        switch(max)
        {
            case 1:
                if(test[i].type == 1)
                    correct ++;
                else 
                    wrong++;
                break;
            
            case 2:
                if(test[i].type == 2)
                    correct ++;
                else 
                    wrong++;
                break;

            case 3:
                if(test[i].type == 3)
                    correct ++;
                else 
                    wrong++;
                break;
        }
    }
    return correct;
}

//To print dataset
void print(vector<first> v)
{
    for(int i=0; i<v.size(); i++)
    {
        for(int j=0; j<4; j++)
        {
            cout <<v[i].arr[j] <<" ";
        }
        cout <<v[i].type;
        cout <<endl;
    }
    cout <<endl;
}

int main()
{
    vector<first> train, test;
    split(train, test);

    cout <<"\nTrain data is: " <<endl;
    print(train);
    cout <<"\nTest data is: " <<endl;
    print(test);

    int max = 0;
    int K;
    for(int k=1; k<100; k = k+2)
    {
        int correct = KNN(k,test,train);
        float per = (correct*100)/test.size();

        if(per > max)
        {
            max = per;
            K = k;
        }

        cout <<"Accuracy of k = " <<k <<" is " <<per <<"%" <<endl;
    }

    cout <<"\nMaximum accuracy is at k = " <<K <<" and the accuracy is = " <<max <<"%" <<endl;
    
}
